crates/model/src/pipe.rs

```diff
@@ -71,26 +74,20 @@ impl Pipe {
     }
 
     pub fn tick(&mut self, size: (u16, u16), rng: &mut Rng, turn_chance: f32) -> InScreenBounds {
-        let InScreenBounds(in_screen_bounds) =
-            self.pos.move_in(self.dirs[self.dirs.len() - 1], size);
+        let InScreenBounds(in_screen_bounds) = self.pos.move_in(self.dir.current(), size);
 
         if !in_screen_bounds {
             return InScreenBounds(false);
         }
 
-        self.dirs.push(*self.dirs.last().unwrap());
-        self.dirs.last_mut().unwrap().maybe_turn(rng, turn_chance);
+        self.dir.update(|dir| dir.maybe_turn(rng, turn_chance));
 
         InScreenBounds(true)
     }
 
     pub fn to_char(&self) -> char {
-        let dir = self.dirs[self.dirs.len() - 1];
```

Comment:
I didn’t even realise how messy this code was before I went to delete it. A mixture of `self.dirs[self.dirs.len() - 1]`, `self.dirs.last()` and `self.dirs.last_mut()` when you could replace them all with a single `let last_dir = self.dirs.last_mut()` at the top.